src/Bicep.Core/Semantics/ParamsSemanticModel.cs

```diff
@@ -67,24 +140,40 @@ private IEnumerable<IDiagnostic> GetAdditionalSemanticDiagnostics()
             return diagnosticWriter.GetDiagnostics();
         }
         
-        private void GetParameterMismatchDiagnostics(IDiagnosticWriter diagnosticWriter, ImmutableArray<Metadata.ParameterMetadata> parameters, IEnumerable<ParameterAssignmentSyntax> parameterAssignments)
+        private void GetParameterMismatchDiagnostics(IDiagnosticWriter diagnosticWriter, IEnumerable<ParameterDeclarationSyntax> parameters, IEnumerable<ParameterAssignmentSyntax> parameterAssignments)
         {
-            var parametersDict = parameters.ToDictionary(x => x.Name, LanguageConstants.IdentifierComparer);
-            var parametersAssignmentsDict = parameterAssignments.ToDictionary(x => x.Name.IdentifierName, LanguageConstants.IdentifierComparer);
-            
-            // parameters that are declared but not assigned
-            var missingRequiredParams = parametersDict.Values
-                .Where(x => x.IsRequired)
-                .Where(x => !parametersAssignmentsDict.ContainsKey(x.Name))
-                .OrderBy(x => x.Name)
-                .Select(x => x.Name)
-                .ToArray();
             // parameters that are assigned but not declared
-            var missingAssignedParams = parametersAssignmentsDict
-                .Where(x => !parametersDict.ContainsKey(x.Key))
-                .Select(x => x.Value)
-                .OrderBy(x => x.Name.IdentifierName);
+            var missingAssignedParams = new List<ParameterAssignmentSyntax>();
+
+            foreach (var parameterAssignment in parameterAssignments)
+            {
+                if (this.ParamBinder.GetSymbolInfo(parameterAssignment) is ParameterAssignmentSymbol symbol && GetParameterDeclaration(symbol) is null)
+                {
+                    missingAssignedParams.Add(parameterAssignment);
+                }
+            }
```

Comment:
`missingAssignedParams` can be an IEnumerable and you can simplify the code with a Linq query, for example:

```C#
var missingAssignedParams = parameterAssignments
    .Where(x => this.ParamBinder.GetSymbolInfo(parameterAssignment) is ParameterAssignmentSymbol symbol && this.TryGetParameterDeclaration(symbol) is null);
```